crates/pallet-native-to-evm-currency-swap/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+    impl<T: Config> Pallet<T> {
+        /// Swap balances.
+        #[pallet::call_index(0)]
+        pub fn swap(
+            origin: OriginFor<T>,
+            to: T::EvmAccountId,
+            #[pallet::compact] amount: NativeBalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            with_storage_layer(move || {
+                Self::do_swap(who, to, amount, Preservation::Expendable)?;
+
+                Ok(())
+            })
+        }
+
+        /// Same as the swap call, but with a check that the swap will not kill the origin account.
+        #[pallet::call_index(1)]
+        pub fn swap_keep_alive(
+            origin: OriginFor<T>,
+            to: T::EvmAccountId,
+            #[pallet::compact] amount: NativeBalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            with_storage_layer(move || {
+                Self::do_swap(who, to, amount, Preservation::Preserve)?;
+
+                Ok(())
+            })
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// General swap balances implementation.
+        pub fn do_swap(
+            who: T::AccountId,
+            to: T::EvmAccountId,
+            amount: NativeBalanceOf<T>,
+            preservation: Preservation,
+        ) -> DispatchResult {
+            let estimated_swapped_balance = T::BalanceConverter::convert(amount);
+            T::EvmCurrency::can_deposit(&to, estimated_swapped_balance, Provenance::Extant)
+                .into_result()?;
+
+            T::NativeCurrency::transfer(&who, &T::PotNativeBridge::get(), amount, preservation)?;
+
+            let evm_balance_to_be_deposited: u128 =
+                estimated_swapped_balance.unique_saturated_into();
+
+            let transaction = pallet_ethereum::Transaction::EIP1559(ethereum::EIP1559Transaction {
+                chain_id: <T as pallet_evm::Config>::ChainId::get(),
+                nonce: pallet_evm::Pallet::<T>::account_basic(&T::PotEvmBridge::get().into())
+                    .0
+                    .nonce,
+                max_priority_fee_per_gas: 0.into(),
+                max_fee_per_gas: 0.into(),
+                gas_limit: 21000.into(), // simple transfer
+                action: ethereum::TransactionAction::Call(to.clone().into()),
+                value: U256::from(evm_balance_to_be_deposited),
+                input: Default::default(),
+                access_list: Default::default(),
+                odd_y_parity: false,
+                r: Default::default(),
+                s: Default::default(),
+            });
+
+            let evm_trx_hash = transaction.hash();
+
+            let _post_info = pallet_ethereum::ValidatedTransaction::<T>::apply(
+                T::PotEvmBridge::get().into(),
+                transaction,
+            )
+            .map_err(|dispatch_error_with_post_info| dispatch_error_with_post_info.error)?;
```

Comment:
Should we take into account `post_info` as `apply` return `DispatchResultWithPostInfo`, adjust it with native transfer info and return `DispatchResultWithPostInfo` instead of `DispatchResult` at the calls ?

Reply:
I don't know. It looks like the bench would take care of this, no?

Reply:
In short: we'd need to bench something, and it will depend on our bench structure. As the code is right now, the bench will cover all of it, and there's no need to adjust the post info.